x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportInferTrainedModelDeploymentAction.java

```diff
@@ -42,15 +45,29 @@ protected void doExecute(Task task, InferTrainedModelDeploymentAction.Request re
         String deploymentId = request.getDeploymentId();
         // We need to check whether there is at least an assigned task here, otherwise we cannot redirect to the
         // node running the job task.
-        PersistentTasksCustomMetadata tasks = clusterService.state().getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
-        PersistentTasksCustomMetadata.PersistentTask<?> deploymentTask = MlTasks.getTrainedModelDeploymentTask(deploymentId, tasks);
-        if (deploymentTask == null || deploymentTask.isAssigned() == false) {
+        TrainedModelAllocation allocation = TrainedModelAllocationMetadata
+            .allocationForModelId(clusterService.state(), request.getDeploymentId())
+            .orElse(null);
+        if (allocation == null) {
             String message = "Cannot perform requested action because deployment [" + deploymentId + "] is not started";
             listener.onFailure(ExceptionsHelper.conflictStatusException(message));
-        } else {
-            request.setNodes(deploymentTask.getExecutorNode());
-            super.doExecute(task, request, listener);
+            return;
+        }
+        String[] randomRunningNode = allocation.getNodeRoutingTable()
+            .entrySet()
+            .stream()
+            .filter(entry -> RoutingState.STARTED.equals(entry.getValue().getState()))
+            .map(Map.Entry::getKey)
+            .toArray(String[]::new);
+        if (randomRunningNode.length == 0) {
+            String message = "Cannot perform requested action because deployment [" + deploymentId + "] is not yet running on any node";
+            listener.onFailure(ExceptionsHelper.conflictStatusException(message));
+            return;
         }
+        // TODO Do better routing for inference calls
+        int nodeIndex = Randomness.get().nextInt(randomRunningNode.length);
+        request.setNodes(randomRunningNode[nodeIndex]);
+        super.doExecute(task, request, listener);
```

Comment:
because we do tasks, and still use the object `TrainedModelDeploymentTask` all we need to do is set what node we care about and the internal routing does the rest (filtering to the correct node, finding the right task, allowing us to infer against it).